### PR TITLE
[Add] `transformers` text and image models

### PR DESCRIPTION
**Pull request by ariG23498**
This PR adds the provision of evaluating models using `transformers`.

One can start a run by the following command
```sh
python clip_benchmark/cli.py eval \
--model_type "transformers" \
--model "openai" \
--pretrained "clip-vit-base-patch32" \
--dataset "imagenet1k" \
--seed 40 \
--task "zeroshot_classification" \
--output "result_{model}_{pretrained_full_path}_{task}_{dataset}_{language}.json" \
--dataset_root "{dataset}" \
--model_cache_dir "."
```

### CODE REVIEW

File: clip_benchmark/models/transformers_clip.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        self.model = model
+    
+    def encode_text(self, text):
+        return self.model.get_text_features(**text)
+
+    def encode_image(self, image):
+        return self.model.get_image_features(image["pixel_values"].squeeze(1))
+
+def load_transformers_clip(model_name, pretrained, cache_dir, device):
+    ckpt = f"{model_name}/{pretrained}"
```
Comment (qubvel) [inline clip_benchmark/models/transformers_clip.py:17]: `ckpt = f"{model_name}/{pretrained}"` may confusing, it's better to provide `model_name` as checkpoint on the hub, and hardcode `pretrained` as `True` IMO. Otherwise it's going to be like

model_name = "openai"
pretrained = "clip-..."

File: clip_benchmark/models/transformers_clip.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+class TransformerWrapper(nn.Module):
+    def __init__(self, model):
+        super().__init__()
+        self.model = model
+    
+    def encode_text(self, text):
+        return self.model.get_text_features(**text)
+
+    def encode_image(self, image):
+        return self.model.get_image_features(image["pixel_values"].squeeze(1))
```
Comment (qubvel) [inline clip_benchmark/models/transformers_clip.py:14]: Interestingly, why should we remove 1-st dim?

File: clip_benchmark/models/transformers_clip.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+def load_transformers_clip(model_name, pretrained, cache_dir, device):
+    ckpt = f"{model_name}/{pretrained}"
+    model = AutoModel.from_pretrained(ckpt, cache_dir=cache_dir, device_map=device)
+    model = TransformerWrapper(model)
+    processor = AutoProcessor.from_pretrained(ckpt)
+
+    transforms = partial(processor.image_processor, return_tensors="pt") 
+    tokenizer = partial(processor.tokenizer, return_tensors="pt", padding="max_length")
```
Comment (qubvel) [inline clip_benchmark/models/transformers_clip.py:23]: AFAIR it might be good to be able to pass additional args for the tokenizer, e.g. for siglip we should specify 
`padding="max_length", max_length=64`

File: clip_benchmark/models/transformers_clip.py (hunk 2 above)
Comment (ariG23498) [reply clip_benchmark/models/transformers_clip.py:14]: I wanted to brain storm on this bit, thanks for asking this question.
The load function should return a model, transforms, and a tokenizer. In the load function I have written it returns the model, the image processor, and the tokenizer.

The transform is used in the collation function while building the dataloader, which is where it adds another dimension to the tensors. So I noticed that the `images["pixel_values].shape=(b, 1, c, h, w)`.

Is there a way I could extract the `transform` function from an `image_processor`?

File: clip_benchmark/models/transformers_clip.py (hunk 3 above)
Comment (ariG23498) [reply clip_benchmark/models/transformers_clip.py:23]: I would have loved to pass extra parameters too, but there is no way to pass extra parameters to the `load` function. The only way we pass parameters is by arguments provided to the script.

File: clip_benchmark/models/transformers_clip.py (hunk 1 above)
Comment (ariG23498) [reply clip_benchmark/models/transformers_clip.py:17]: I had to choose this option for better verbosity.

https://github.com/LAION-AI/CLIP_benchmark/blob/a23028288ed914efb31c5ff32a1e702123cbd0ff/clip_benchmark/cli.py#L247

**Comment by ariG23498** [comment]
Worth mentioning with the current integration to `transformers` I ran zero shot evaluation with open_clip siglip and transformers siglip

```
python clip_benchmark/cli.py eval \
--model_type "open_clip" \
--model "ViT-SO400M-14-SigLIP-384" \
--pretrained "webli" \
--dataset "imagenet1k" \
--seed 40 \
--task "zeroshot_classification" \
--output "result_{model}_{pretrained_full_path}_{task}_{dataset}_{language}.json" \
--dataset_root "{dataset}" \
--model_cache_dir "."

python clip_benchmark/cli.py eval \
--model_type "transformers" \
--model "google" \
--pretrained "siglip-so400m-patch14-384" \
--dataset "imagenet1k" \
--seed 40 \
--task "zeroshot_classification" \
--output "result_{model}_{pretrained_full_path}_{task}_{dataset}_{language}.json" \
--dataset_root "{dataset}" \
--model_cache_dir "."
```

Results:
|    | dataset    | model                    | pretrained                | task                    | metrics                                                                                               | language   |
|---:|:-----------|:-------------------------|:--------------------------|:------------------------|:------------------------------------------------------------------------------------------------------|:-----------|
|  0 | imagenet1k | google                   | siglip-so400m-patch14-384 | zeroshot_classification | {'acc1': 0.8306600000000001, 'acc5': 0.9703600000000001, 'mean_per_class_recall': 0.8310000000000001} | en         |
|  1 | imagenet1k | ViT-SO400M-14-SigLIP-384 | webli                     | zeroshot_classification | {'acc1': 0.83084, 'acc5': 0.9703600000000001, 'mean_per_class_recall': 0.8310200000000001}            | en         |

File: clip_benchmark/models/transformers_clip.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+def load_transformers_clip(model_name, pretrained, cache_dir, device):
+    ckpt = f"{model_name}/{pretrained}"
+    model = AutoModel.from_pretrained(ckpt, cache_dir=cache_dir, device_map=device)
+    model = TransformerWrapper(model)
+    
+    processor = AutoProcessor.from_pretrained(ckpt)
+    transforms = partial(processor.image_processor.preprocess, return_tensors="pt")
+    tokenizer = partial(
+        processor.tokenizer, return_tensors="pt", padding="max_length", max_length=64
+    )
```
Comment (qubvel) [inline clip_benchmark/models/transformers_clip.py:28]: We might include some additional logic into the `Processor` class, I feel this way it would be more safe
```suggestion
    transforms = lambda image: processor(images=image, return_tensors="pt")
    tokenizer = lambda text: processor(text=text, padding="max_length", max_length=64, return_tensors="pt")
```

File: clip_benchmark/models/transformers_clip.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import torch
+from torch import nn
+from transformers import AutoModel, AutoProcessor
+from functools import partial
```
Comment (qubvel) [inline clip_benchmark/models/transformers_clip.py:4]:
```suggestion
```

File: clip_benchmark/models/transformers_clip.py (hunk 4 above)
Comment (ariG23498) [reply clip_benchmark/models/transformers_clip.py:28]: TIL! This is definitely a better way.
